lms/services/grouping.py

```diff
@@ -0,0 +1,68 @@
+from lms.models import CanvasSection, Grouping
+from lms.models._hashed_id import hashed_id
+
+
+class GroupingService:
+    def __init__(self, db, application_instance_service, course_service):
+        self._db = db
+        self._application_instance = application_instance_service.get()
+        self._course_service = course_service
+
+    def upsert(self, grouping):
+        db_grouping = (
+            self._db.query(Grouping)
+            .filter_by(
+                application_instance=grouping.application_instance,
+                authority_provided_id=grouping.authority_provided_id,
+                type=grouping.type,
+            )
+            .one_or_none()
```

Comment:
I don't think two grouping of different types can have the same `authority_provided_id`, otherwise their groups in the h DB would clash, so the `type` might be unnecessary in this filter.

Reply:
Yeah there's a unique constraint: https://github.com/hypothesis/lms/blob/88698f5ce32e87a59f66d9c87054885bfc02d119/lms/models/grouping.py#L15